### ISSUE TITLE
Change default Python optimizer for GPflow?

### ISSUE THREAD

**Issue by alexggmatthews** [GPflow/GPflow#188]
At the moment we use lbfgs_b from scipy as the default optimizer in GPflow. 

Particularly with #185 and using float32 and float64 we have some anecdotal evidence that lbfgs_b is aggressively finding numerical instabilities in the algorithm where other optimizers wouldn't.

Rasmussen's minimize.m springs to mind as an algorithm with good performance and numerical stability. We might of course consider other methods. 

I know @markvdw is pretty knowledgeable on this topic so I'm assigning this issue to him for comments.

**Comment by markvdw** [comment]
Not sure to what extent this is a satisfying answer, but my experience is that the scipy optimisers can be quite brittle. Both in terms of aggressiveness and in terms of giving up. Carl's "minimize.m" is currently an implementation of bfgs, only with a bunch of sensible modifications to allow it to keep chugging along, for example switching to gradient descent for a few iterations where scipy would just give up.

My only real recommendation would be that running CG or gradient descent for the first few iterations is not a bad idea to counteract the aggressiveness that @jameshensman talks about.

**Comment by jameshensman** [comment]
How much work would it be to get minimize.m going in GPflow? Seems like a better default option than the very brittle scipy.lbfgs.

**Comment by markvdw** [comment]
I'm actually working with Carl on a "proper solution" to some other problems in minimize at the moment. And I do plan to make a Python implementation.

The issue with "proper solution" is that it may take some time (on the scale of a month or so), and that I don't really want to port a halfway solution if I'm just going to re-do it later.

**Comment by jameshensman** [comment]
Okay! That sounds  great, I look forward to having a super optimizer in GPflow in a month :)

**Comment by markvdw** [comment]
Come to think of it, it would be nice to collect a set of cases where the scipy optimiser gets stuck, or goes to a numerically unstable region, both to test the improved minimize, and to quantify how much of a problem it is in GPflow (and how much it has improved).

If anybody wants to send me a situation like this, and doesn't mind it being used in our paper, get in touch!

**Comment by mathDR** [comment]
I don't know how much you guys are interested in this area, but William W.
Hager at the University of Florida has a lot of research software regarding
numerical optimization:  http://users.clas.ufl.edu/hager/

You can pick and choose, but his conjugate gradient code is very good.  He
also has a bunch of sparse Cholesky solvers and the like.

Again, I don't know how deep down the rabbit hole you want to go in this
area...

Dan

On Wed, Aug 31, 2016 at 9:30 AM, Mark van der Wilk <notifications@github.com

**Comment by kacreel** [comment]
Hi GPflow developers, thank you for a wonderful library.  It is forming the backbone of my sea level change modeling, and I really appreciate the work you do.  

An optimization question followup:  when using GPR on a 3D kernel I consistently get stuck in local minima with the scipy.optimize.minimize optimizers.  Has there been any discussion about including a global minimizing function such as scipy's basinhopping or dual annealing?

**Comment by rcameronc** [comment]
Hi GPflow developers, thank you for a wonderful library. It is forming the backbone of my sea level change modeling, and I really appreciate the work you do.

An optimization question followup: when using GPR on a 3D kernel I consistently get stuck in local minima with the scipy.optimize.minimize optimizers. Has there been any discussion about including a global minimizing function such as scipy's basinhopping or dual annealing?

**Comment by st--** [comment]
Hi @rcameronc ! thanks for the positive feedback, really good to hear after all the effort and time we've put into it :) and glad it's being put to good use! (If any of your work is public, we'd love to hear about it - e.g. feel free to make a pull request to add your project to the README.md file).

We've not thought much about using non-gradient based methods - though especially with tensorflow 2.0 this should be a lot easier than before. Hard to say anything concrete without more details.

What do you mean by 3D kernel? how do you initialize your kernel hyperparameters?

**Comment by awav** [comment]
@rcameronc,

> Has there been any discussion about including a global minimizing function such as scipy's basinhopping or dual annealing?

No, we didn't discuss these optimisation methods. At first glance, they don't look scary and I think it should be relatively easy to introduce them into our scipy interface.

**Comment by rcameronc** [comment]
Hi @st--, I will definitely make a pull request with my project once I hammer out the kinks and submit it for publication.  Thank you for the invitation.  

re: 3D kernel, I mean that I make a composite kernel from a 2 dimensional matern32 kernel operating on lat/lon multiplied by a 1-D matern32 kernel on time plus a white noise kernel. 

The bounded hyperparameters are initialized via this function: 

' ' ' 
def bounded_parameter(low, high, param):
    """Make parameter tfp Parameter with optimization bounds."""

    sigmoid = tfb.Sigmoid(low=tf.cast(low, tf.float64), 
                          high=tf.cast(high, tf.float64),
                         name='sigmoid')
    parameter = gpf.Parameter(param, transform=sigmoid, dtype=tf.float64)
    return parameter
' ' ' 
which I call like this: 
k4 = gpf.kernels.Matern32(active_dims=[2]) 
k4.lengthscale = bounded_parameter(1, 30000, 186)

 I'll try to simplify it into a MWE.  

@awav  I agree, it seems like it would be relatively straightforward to include them in the scipy interface.  Would that mean adding basinhopping and/or dual_annealing as new classes within [gpflow/optimizers/scipy](https://github.com/GPflow/GPflow/blob/develop/gpflow/optimizers/scipy.py)?
